crates/sui-framework/sources/math.move

```diff
@@ -64,4 +64,16 @@ module sui::math {
 
         (res as u64)
     }
+    
+    /// Return the value of a base raised to a power
+    public fun pow(base: u64, exponent: u64): u64 {    
+        let res = 1;
+        let i = 0;
+        while (i < exponent) {
+            res = (res * base);
+            i = i + 1 ;
+        };
```

Comment:
Let's use the logarithmic exponentiation algorithm, which I *think* (I'm offering this suggestion without building or testing it) would look like this:

```suggestion
        let res = 1;
        while (exponent >= 1) {
            if (exponent % 2 == 0) {
                base = base * base;
                exponent = exponent / 2;
            } else {
                res = res * base;
                exponent = exponent - 1;
            }
        };
```

Reply:
Thank you @amnn,

How many cases would you expect the tests to cover? Would this be sufficient? Too many assertions?

```
#[test]
fun test_pow() {
    assert!(nft::pow(1, 1) == 1, 0);
    assert!(nft::pow(2, 10) == 1024, 0);
    assert!(nft::pow(10, 6) == 1000000, 0);
}
```

I'm not sure how to write the overflow tests. Is it acceptable to just catch a panic from the `pow` function? Something like this...

```
#[test]
#[expected_failure]
fun test_pow_overflow() {
    nft::pow(10, 100);
}
```

or is there some way to specifically check for an overflow/arithmetic error (whatever it would be. I'm not sure)?

I apologise for the noob questions, but I'm not very experienced writing test.

Reply:
No, these are great questions.  That kind of testing looks good to me (I would also include a test for taking to the power of 0), and yes overflows will cause a panic, so `expected_failure` will work.  You can provide the exact abort code (which you can find by getting rid of the `expected_failure` annotation and then running the test):

```
#[expected_failure(abort_code = _)]
```

Reply:
Looking at the tests you have now, I would add a couple more of the edge cases that you mentioned in the comment, and I would also add a test case to make sure that `pow128` works for values that would have overflowed `pow`.

Reply:
```
┌── test_pow_overflow ──────
│ ITE: An unknown error was reported. Location: error[E11001]: test failure
│    ┌─ ./sources/nft.move:66:29
│    │
│ 62 │     public fun pow(base: u64, exponent: u8): u64 {    
│    │                --- In this function in 0x0::nft
│    ·
│ 66 │                 base = base * base;
│    │                             ^
│ 
│ 
│ VMError (if there is one): VMError {
│     major_status: ARITHMETIC_ERROR,
│     sub_status: None,
│     message: None,
│     exec_state: None,
│     location: Module(
│         ModuleId {
│             address: 0000000000000000000000000000000000000000,
│             name: Identifier(
│                 "nft",
│             ),
│         },
│     ),
│     indices: [],
│     offsets: [
│         (
│             FunctionDefinitionIndex(1),
│             16,
│         ),
│     ],
│ }
└──────────────────
```

This is the error when I remove `expected_failure`. How would I determine the correct abort code?

Reply:
Ah thanks for exploring this.  It looks like it's not an `abort` so the system for testing a specific abort code won't work here.  That's okay, we can stick with just `expected_failure`.